src/main/java/org/wildfly/security/auth/provider/oauth2/OAuth2SecurityRealm.java

```
@@ -0,0 +1,270 @@
+                this.evidence = (BearerTokenEvidence) evidence;
+            } else {
+                this.evidence = null;
+            }
+        }
+        @Override
+        public Principal getRealmIdentityPrincipal() {
+            try {
+                if (exists()) {
+                    return new NamePrincipal(getClaims().getString(configuration.principalClaimName));
+                }
+            } catch (Exception e) {
+                throw ElytronMessages.log.oauth2RealmFailedToObtainPrincipal(e);
+            }
+            return null;
+        }
+
+        @Override
+        public boolean verifyEvidence(Evidence evidence) throws RealmUnavailableException {
+            return isValidToken(introspectToken());
+        }
+
+        @Override
+        public boolean exists() throws RealmUnavailableException {
+            return getClaims() != null;
+        }
+
+        @Override
+        public AuthorizationIdentity getAuthorizationIdentity() throws RealmUnavailableException {
+            if (exists()) {
+                return new AuthorizationIdentity() {
+                    private Attributes attributes;
+
+                    @Override
+                    public Attributes getAttributes() {
+                        if (this.attributes == null) {
+                            this.attributes = toAttributes(claims);
+                        }
+
+                        return this.attributes;
+                    }
+                };
+            }
+
+            return null;
+        }
+
+        @Override
+        public SupportLevel getCredentialAcquireSupport(Class<? extends Credential> credentialType, String algorithmName) throws RealmUnavailableException {
+            return SupportLevel.UNSUPPORTED;
+        }
+
+        @Override
+        public <C extends Credential> C getCredential(Class<C> credentialType) throws RealmUnavailableException {
+            throw new RealmUnavailableException("Unsupported operation");
```

We probably need to update the javadoc on RealmIdentity to make it a bit clearer but at this point null should be silently returned - RealmUnavailableException is more about transient issues e.g. being unable to connect to a remote server.